### DISCUSSION TITLE
Proxies

### DISCUSSION THREAD

**Discussion by Pingo-v0** [kaliiiiiiiiii/Selenium-Profiles#6] [Q&A] [answered]
Looks great ! (seems to work fine while many other solutions had failed -thanks)

Would there be a way to set the proxies when creating the driver instance ?

**Comment by ttlns** [comment]
Thank you!

Adding proxies should be possible as following:
```python
from selenium_profiles.driver import driver as mydriver
from selenium_profiles.profiles import profiles
from selenium.webdriver.common.by import By  # locate elements

mydriver = mydriver()
profile = profiles.Windows() # or .Android


PROXY = "23.23.23.23:3128" # IP:PORT or HOST:PORT
PROXY_METHOD = "socks5://" # optional, default is socks5

# add to profile
profile["options"]["browser"].update({"proxy":PROXY, "proxy_method":PROXY_METHOD})


driver = mydriver.start(profile, uc_driver=False) 

# get url
driver.get('https://browserleaks.com/client-hints')  # test client hints

input("Press ENTER to exit: ")
driver.quit()  # Execute on the End!
```

For other options in the profile, have a look at [Example-Profile](https://github.com/kaliiiiiiiiii/Selenium-Profiles#profiles)

Didn't test it yet tho, but in theory it should work like that:) Feel free to tell me if it works as intended.

**Reply by Pingo-v0** [reply]
Mmh.. it doesn't seem to work : (

Maybe because the proxies provided by the service I use ('brightdata') contain an authentification token.
Here are their format :
{
        'http': 'http://user:pass@zproxy.lum-superproxy.io:22225',
        'https': 'https://user:pass@zproxy.lum-superproxy.io:22225'
}

I tried :
profiles['browser']['proxy']="user:pass@zproxy.lum-superproxy.io:22225"
profiles['browser']['proxy_method']="http://"
But it didnt work

Any idea whether there is a workaround to make it work ?

PS : here is the open-api I use to check whether the proxy works : http://lumtest.com/myip.json
It does work when using "requests.get(url, proxies=xxx) but not when setting the library with profile['browser']['proxy']..

**Reply by ttlns** [reply]
Oh I think I see the problem:
the `proxy` and `proxy` keys gotta be in `profile["options"]["browser"]` and not `profile["browser"]`

I might need to implement some propper error-handling on invalid keys

**Reply by Pingo-v0** [reply]
Thanks for the quick answer, So no workaround so far ?

I tried both :

    prof = profiles.Windows()
    options = {
        'browser': {
            'proxy': PROXY,
            'proxy_method': PROXY_METHOD}}
    prof['options'] = options

and :

    prof = profiles.Windows()
    bb =  {
        'proxy': PROXY,
        'proxy_method': PROXY_METHOD}
    prof["chromeoptions"]['browser']=bb

But doesn't seem to work, should it ?

**Reply by ttlns** [reply]
Oh well that's weird, I'll have a look👍

**Reply by Pingo-v0** [reply]
Thanks

**Reply by ttlns** [reply]
http proxy without credentials works for me as following:
```python
PROXY = "193.123.98.126:8080" # IP:PORT or HOST:PORT
PROXY_METHOD = "http://" # optional, default is socks5

# add to profile
profile["options"]["browser"].update({"proxy":PROXY, "proxy_method":PROXY_METHOD})


driver = mydriver.start(profile, uc_driver=False)

driver.get("http://lumtest.com/myip.json")
```

Note: Proxies used from https://free-proxy-list.net/

It seems, like chrome doesn't suport specifying proxy credentials (see https://stackoverflow.com/questions/55582136/how-to-set-proxy-with-authentication-in-selenium-chromedriver-python)


Solutions mentioned are:
- use a [local proxy](https://stackoverflow.com/a/71413054/20443541) (without authentification) which redirects your traffic to the desired proxy
- Use [pyautogui](https://stackoverflow.com/a/56353897/20443541) to authentificate when a authentification dialog pops up, and leaving out the credentials in the profile (Doesn't work multithreaded or headless)
- Use a [extension](https://stackoverflow.com/a/55582859/20443541) I think this is the ___best method___, but needs some modification and development. I might implement that in the future, but don't have that much time atm for that. A example script (wothout using selenium-profiles would be at https://github.com/Smartproxy/Selenium-proxy-authentication)

**Reply by Pingo-v0** [reply]
The extension option looks great.
Without taking all your time, would there be a simple way **for me** (is it possible for me to do it, not being a specialist of python libraries) to change your driver.py source file as follow (and would it work ?):

- add a parameter to the **start** function :
 `start(self, profile: dict, uc_driver: bool = False, proxy:dict = None)`
where `proxy = {"PROXY_HOST":xxx, "PROXY_PORT":xxx, "PROXY_USER":xxx, "PROXY_PASS":xxx}`
 
- inject all the code described in the [extension](https://stackoverflow.com/a/55582859/20443541) mentionned in the driver.py file
(mainly the get_chromedriver function which would return a "chrome_options" object instead of a driver)
```
def get_chromeoptions(proxy:dict):
    ...
    creation of manifest_json, background_js, etc.
    ...
    return chrome_options
```

- and directly add the chrome_options parameter in line 84 when creating the instance of webdriver.Chrome (chrome_options being generated thanks to the additionnal parameters provided in the start function AND thanks to the get_chromedriver function from the code mentionned above)
`self.driver = webdriver.Chrome(options=self.options, chrome_options=get_chromeoptions(proxy))`

**Reply by ttlns** [reply]
Well yes I think that should work, just not using the `chrome_options` argument.
Just add:
`self.options.options.add_extension("full_path_to_extension.zip")` before line [84](https://github.com/kaliiiiiiiiii/Selenium-Profiles/blob/master/src/selenium_profiles/driver.py#L84).
if you don't use `self.options` as the options for the driver in line 84, elsewise the profile["options"] arguments wont have an effect anymore.

That will only work without uc_driver tho

**Reply by Pingo-v0** [reply]
So last question : once I've installed your library using : `pip install selenium-profiles`
Where can I find the sources files ? (not found in my virtual environment .venv folder...)

**Reply by ttlns** [reply]
think they should be in venv/lib/packages/selenium_profiles or something like that

**Reply by Pingo-v0** [reply]
(That was a silly question..)

Thanks for all the answers !

**Reply by Pingo-v0** [reply]
In case you see what's wrong (don't hesitate to stop my questions, which may go on for a while if you don't since I'm trying to make it work not knowing how much work is remaining)

1. self.options.options.add_extension("full_path_to_extension.zip") returned the following error :
`AttributeError: 'Options' object has no attribute 'options'`
=> I assumed you meant self.options.add_extension("full_path_to_extension.zip")

2. Then a new error popped up :

```
driver = mydriver.start(prof, uc_driver=False, proxy=proxy)  # or .Android
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium_profiles/driver.py", line 174, in start
self.driver = webdriver.Chrome(options=self.options)  # start selenium webdriver
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/chrome/webdriver.py", line 81, in __init__
super().__init__(
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/chromium/webdriver.py", line 106, in __init__
super().__init__(
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 288, in __init__
self.start_session(capabilities, browser_profile)
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 381, in start_session
response = self.execute(Command.NEW_SESSION, parameters)
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 444, in execute
self.error_handler.check_response(response)
File "/home/zephyr/Git/api_token/.venv/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 249, in check_response
raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: failed to wait for extension background page to load: chrome-extension://jhdijllimahpfmeiniakbilgljppfmlp/_generated_background_page.html
from unknown error: page could not be found: chrome-extension://jhdijllimahpfmeiniakbilgljppfmlp/_generated_background_page.html

```

Any idea ?

PS : this guy seems to have had the same issue 
https://www.solveforum.com/forums/threads/solved-python-selenium-headless-mode-not-working-with-proxy.2108486/

**Reply by ttlns** [reply]
Oh yes, exactly, sorry that was a typo.
Soo I think the extension got loaded, but then produced en error.  Where do you specify that chome should headless? In the profile? Is it even headless?

If you specify it using '--headless' as an argument, you might try out using '--headless=chrome' or without headless.

Oh and btw you can ask me as much as you want. I'l just aswer when I do have time anyways. I mean you help me to develop & test

**Reply by Pingo-v0** [reply]
Here is what I did :
```
prof = profiles.Windows()
prof['browser']['headless']=True
```
(without the proxy, it does work)

**Reply by ttlns** [reply]
Good, that should automatically support extensions in headless

**Comment by ttlns** [comment, marked as answer]
### Proxy-authentication
see [documentation](https://github.com/kaliiiiiiiiii/Selenium-Profiles#set-proxies-dynamically-or-with-options)

EDIT
- temp_dir as argument added @Pingo-v0 Version >= 2.2.5.4)
- new syntax for starting driver and profile >= 2.2.6
- new syntax & dynamic proxies supported >= 2.2.7

**Reply by ttlns** [reply]
@PAJeantaud  Please tell me, if that works in your case.

**Reply by Pingo-v0** [reply]
Hi @kaliiiiiiiiii 
I had not seen your last post, thanks a lot for the changes made ! -looks promising -fingers crossed. I'll test it tomorrow (day spent on front-end dev issues..) and will of course let you know the results (this topic being crucial to me, no chance I'll give up before having tried as hard as possible)

**Reply by oldwizard1010** [reply]
(Hi from stackoverflow)
It says, selenium-profiles doesn't support linux yet, if it's true, i'm afraid i can't use it(
But i already looked here and probably will try it soon, so anyway, thanks)

**Reply by ttlns** [reply]
Oh just not officially. If you install Chrome and chromedriver in /bin, it should work I think.
Feel free to open a issue In case it doesn't work//error-message.

If you are able to run Selenium, then Selenium-profiles should also work.
What code did you use?

**Comment by Pingo-v0** [comment]
Back(end) in the game :

**1. Minor point** :
`profiles.Windows()['browser']['headless']=True` is no longer working, to be replaced by the following ?
`profiles.Windows()['options']['browser']['headless']=True`

**2. Proxy auth doesn't work so far** (I may have done somthing wrong, I'm looking more closely at it) since
`driver.get("http://lumtest.com/myip.json")` returns the same value at each iteration... And i've got the following error :
```
/home/zephyr/Git/webscrapping_test/.venv/lib/python3.8/site-packages/selenium_profiles/scripts/cdp_tools.py:19: UserWarning: disabling emulation not supported
  warnings.warn('disabling emulation not supported')
```

**Reply by ttlns** [reply]
1. Exactly, the profile object changed. https://github.com/kaliiiiiiiiii/Selenium-Profiles#profiles should be valid currently
2. This error was related to the previous version and only a warning. Since disabling emulation is supported now, I'll remove the warning.

I'm assuming you tested with [this](https://github.com/kaliiiiiiiiii/Selenium-Profiles/discussions/6#discussioncomment-4704385) script? I only tested it without authentification yet, since I could't find a (woking) free one using authentification. I can check authntification with a local proxy, but then I can't check if there's a different ip.

**Reply by ttlns** [reply]
This configuration: 
```python
host = "zproxy.lum-superproxy.io" # IP
port = 22225
user = "lum-customer-XXXXXXX-zone-residential-country-fr-session-XXXXXXXXXXXXXX"
password = "some_random_password"
scheme='http'

auth_proxy = {"host":host,"port":port,"username":user, "password":password, "scheme":scheme}

profile["options"]["extensions"] = {"auth_proxy":auth_proxy}
```
I can confirm to be working:
```
{
"ip":"193.55.147.180",
"country":"FR",
"asn":{"asnum":2200,"org_name":"Renater"},
"geo":{"city":"Montpellier","region":"OCC","region_name":"Occitanie","postal_code":"34000","latitude":43.6107,"longitude":3.8809,"tz":"Europe/Paris","lum_city":"montpellier","lum_region":"occ"}
}
```
Google wants a captcha (blacklisted IP):
![image](https://user-images.githubusercontent.com/89038706/214039768-5f01dc0e-c3c7-458d-985c-796c80d5fb12.png)
(I'm  currently in cunty "CH")

The https one I can also confirm to be working with the current dev version (install with `pip install https://github.com/kaliiiiiiiiii/Selenium-Profiles/archive/refs/heads/dev.zip`:
- seems to be slower
- no google-captcha needed


Interesting that `requests.get('http://lumtest.com/myip.json', proxies=pp).json()` also works, thanks for that contibution.

Note: The "normal" version (master-branch in this github) doesn't support proxy yet..

**Reply by Pingo-v0** [reply]
It seems to work when headless is not set to True
But to fail when `profile['options']['browser']['headless']=True`
Isn't ? Any idea why.. ?

**Reply by ttlns** [reply]
> It seems to work when headless is not set to True But to fail when `profile['options']['browser']['headless']=True` Isn't ? Any idea why.. ?

Well yes, extensions aren't officially supported by Chrome in headless mode. I heard that the `--headless=Chrome` option (which Selenium-Profiles used by default when specifying `profile['options']['browser']['headless']=True`) works with extensions, but maybe not with proxy.

**Reply by ttlns** [reply]
You clould try by adding the option `--enable-features=NetworkService,NetworkServiceInProcess` (resource: https://stackoverflow.com/a/60398023/20443541)

Passing Chrome-options woks as following:
```python
profile["options"]["option_args"] = ["--my-arg1=some_value", "--my-arg2=some_other_value"]
```

**Reply by Pingo-v0** [reply]
Thanks, just tried, but does not work : (

**Reply by ttlns** [reply]
Well then I don't know another solution atm. Maybe have a look at the other workarounds like setting up a local, unathenticated proxy which redirects to your desired proxy. (kinda like proxy-chain)

**Reply by Pingo-v0** [reply]
Sure, I'm going to dig this issue deeper,
Anyway thanks for the great reactivity (and this library which still solves a point any other I had found would not)

**Reply by ttlns** [reply]
Actually, maybe there's another way using (desired-capabilities](https://chromedriver.chromium.org/capabilities#h.p_ID_52) argument for chrome. But no idea if it supports authentification. 
See https://www.w3.org/TR/webdriver/#proxy

**Reply by Pingo-v0** [reply]
Hi @kaliiiiiiiiii ,
Two more questions about the library:

- does it support POST method ? driver.post(data=json.dumps({...})
- does it handle jvscrpt chllenges ?
![image](https://user-images.githubusercontent.com/28194943/214063998-ee10c568-bb81-45f5-8eaf-e3d2f796bed9.png)

**Reply by Pingo-v0** [reply]
![image](https://user-images.githubusercontent.com/28194943/214064356-0ee54337-4362-4695-baf2-01a3d0380488.png)

**Reply by ttlns** [reply]
Well for post methods, I see 2 options:
- using javascript (see https://stackoverflow.com/a/65309073/20443541)
- using request interception ( I can give you a sample script in case you need it)

For javascript challanges:
- tbh I never really worked with javascript, so not shure what you mean. something like reCaptcha? In that case, ir really depends what type of captcha, so you might provide me with an example//or website? I do have approaches to solve them, but none yet implemented.
- do you mean cloudfare-undetectability? Like how to pass the cloudfare waiting-room?

Note: Maybe move to a new discussion?

**Comment by ttlns** [comment]
Closing as it seems to be resolved
